source/common/http/filter_manager.h

```diff
@@ -837,8 +836,23 @@ class FilterManager : public ScopeTrackedObject,
   virtual StreamInfo::StreamInfo& streamInfo() PURE;
   virtual const StreamInfo::StreamInfo& streamInfo() const PURE;
 
-  // Set up the Encoder/Decoder filter chain.
-  bool createFilterChain();
+  struct CreateFilterChainResult {
+    bool created{};
+    bool upgrade_rejected{};
+  };
+
+  /**
+   * Set up the Encoder/Decoder filter chain.
+   * @param filter_chain_factory the factory to create the filter chain.
+   * @param allow_upgrade_filter_chain whether to allow the creation of an upgrade filter chain.
+   *        This only should be set to true for downstream HTTP filter chain.
+   * @param only_create_if_configured whether to only create the filter chain if it is configured
+   *        explicitly. This only makes sense for upstream HTTP filter chain.
+   *
+   */
+  CreateFilterChainResult createFilterChain(const FilterChainFactory& filter_chain_factory,
```

Comment:
Can you comment a bit more about the motivation for this refactor?
Having a single createFilterChain function where one argument only works for downstream and one for upstream feels ugly.  Why not retain createDownstream/createUpstream and only pass the relevant args?
/wait-any

Reply:
> Can you comment a bit more about the motivation for this refactor?
> /wait-any

The initial motivation is I wanna to refactor the http filter chain to use vector rather than list as the container of filters because the filter chain needn't O(1) modification anyway. And rather than `ActiveStremDecoderFilterPtr`, I want to use `ActiveStremDecoderFilter` directly. This could reduce unnecessary heap allocations.

This refactoring requires us to set filter callbacks at the end of filter chain creation. So I need to ensure the filter chain creation is under the control of `FilterManager`.

And even not for that, make sure all filter chain creation logic in same way is better choice.

> Having a single createFilterChain function where one argument only works for downstream and one for upstream feels ugly.  Why not retain createDownstream/createUpstream and only pass the relevant args?

Yeah, I can make the create filter chain a shared private method and create a wrapper method for upstream filter chain creation.

Reply:
I found I acutally needn't the upgrade option, because we can check whether the downstream stream callbacks is valid or not to determine to try upgrade or not.
Now, I removed it, code is cleaner now.

Reply:
By the way, after a check to the upstream filter chain creation, I think even `only_create_if_configured` is unnecessary. Rather add some specific logic/parameter to the `FilterChainFactory.createFilterChain`, it would be better to use a default FilterChainFactory instance. When no valid filters are provided by both router and cluster, then we can construct a default FilterChainFactory to create the filter chain. (Only contains UpstreamCodecFilter)

This way should be much cleaner.